### ISSUE TITLE
users in container can't write to directories where secondary groups should give write access

### ISSUE THREAD

**Issue by jonathanvaughn** [docker/docker#10252]
Our user logins are provided by SSSD / Kerberos / 389DS, and have various secondary groups provided both via LDAP and /etc/security/group.conf.

Specifically, everyone who logs into the development server is a member of the local apache group (gid 48), via /etc/security/group.conf.

We have several Docker containers set up that run various versions of PHP-FPM (5.4,5.5, etc), and a set of wrappers that allow users to make PHP CLI calls to any particular version (via various magic including using nsenter to become that user inside the container and run the php command they requested). 

We have the various home and web directories bind mounted into the containers, and they generally work fine for most purposes (since typically, the only writes are occurring by web requests which go through PHP-FPM and are performed as the `apache` user/group, and everything is `apache`-writable). 

However, I have recently discovered that while users can write to their own directories with matching group (i.e. `bob` can write to files/directories with group set to `bob`), they can't write to those which they should have secondary group permissions on (i.e. group is set to `apache`, and `bob` is a member of `apache`).

I can reproduce this without our convoluted script magic that allows PHP CLI to be any version through containers, but still using `nsenter` to get into the container:

Entering container:
`[root@dev ~]# nsenter --target PID --mount --uts --ipc --net --pid -- /bin/bash -l`

Becoming the user:
`[root@dev /]# su - bob`

Successfully touching a file in home directory (which is also bind mounted, but owned by bob:bob)
```
[bob@dev ~]$ touch test
```

Going into a test directory set up as 2775 and owned by apache:apache
```
[bob@dev ~]$ cd www/test/apachetest
[bob@dev apachetest]$ touch test2
touch: cannot touch `test2': Permission denied
```

If I repeat the above steps as apache, as the apache user I can touch the file in apachetest directory.

`bob` is a member of apache, even in the container:
```
[root@dev-web ~]# id bob
uid=1005001(bob) gid=1005001(bob) groups=1005001(bob),10(wheel),48(apache),501(devs),504(rvm),1001004(Global Developer Admins), ... 
```

**Comment by jonathanvaughn** [comment]
Actually, might be that `su` isn't using pam_group in the container. D'oh. Will update once I can test this (probably tomorrow). `id` properly displays all groups because it looks them up, but since `su` doesn't actually add the groups the access isn't really there.

**Comment by cpuguy83** [comment]
@jonathanvaughn Need to make sure UID/GID match, not the actual names.
Also, have you tried using `docker exec` instead of nsenter?

**Comment by jonathanvaughn** [comment]
Yes, UID/GID match inside and outside (since that's all pulled from LDAP and/or standardized for CentOS). 

As I said above, I think the immediate problem is that `su` isn't respecting /etc/security/group.conf because it's PAM configuration doesn't use pam_group.

Didn't know about `docker exec`. Looks like this was new in 1.3? I started working on this back when the latest EPEL was 0.9 ... but we have 1.3 from EPEL now. Totally missed that this was added. I'll look into it, it might simplify (and reduce dependencies) the implementation.

**Comment by jonathanvaughn** [comment]
I've verified the cause of the problem at hand is that when root user executes `su`, pam_group.so isn't being called to add secondary groups, but pam_sss.so still is called so they get LDAP groups. When logging in via SSH to a user, using the same included file contents for pam.d, the pam_group.so is called and they get the groups from /etc/security/group.conf in addition to LDAP groups.

As it turns out, this has nothing to do with Docker, so I will close this. We only noticed in a Docker container because we (almost) never `su` from root to a user, and haven't needed secondary groups from /etc/security/group.conf to work in that scenario before. 

I did look at docker exec by the way, and that may simplify our deployment (not needing custom built new enough version of nsenter) but doesn't get around needing to `su - username -c command ...` to a run a command as a user because it executes as root.
